services/rsources/handler.go

```diff
@@ -108,9 +121,133 @@ func (*sourcesHandler) GetFailedRecords(_ context.Context, _ *sql.Tx, _ string,
+	err := sh.doCleanupTables(ctx)
+	if err != nil {
+		return err
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-sh.cleanupTrigger():
+			err := sh.doCleanupTables(ctx)
+			if err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (sh *sourcesHandler) doCleanupTables(ctx context.Context) error {
+	before := time.Now().Add(-config.GetDuration("RSOURCES_STATS_RETENTION", 24, time.Hour))
+	sqlStatement := `delete from "rsources_stats" where job_run_id in (
+		select lastUpdateToJobRunId.job_run_id from 
+			(select job_run_id, max(ts) as mts from "rsources_stats" group by job_run_id) lastUpdateToJobRunId
+		where lastUpdateToJobRunId.mts <= $1
+	)`
+	_, err := sh.localDB.ExecContext(ctx, sqlStatement, before)
+	return err
+}
+
+func (sh *sourcesHandler) readDB() *sql.DB {
+	if sh.sharedDB != nil {
+		return sh.sharedDB
+	}
+	return sh.localDB
+}
+
+func (sh *sourcesHandler) init() error {
+	ctx := context.TODO()
+	if sh.cleanupTrigger == nil {
+		sh.cleanupTrigger = func() <-chan time.Time {
+			return time.After(config.GetDuration("RSOURCES_STATS_CLEANUP_INTERVAL", 1, time.Hour))
+		}
+	}
+	err := setupStatsTable(ctx, sh.localDB, sh.config.LocalHostname)
+	if err != nil {
+		return fmt.Errorf("failed to setup local stats table: %w", err)
+	}
+	if sh.sharedDB != nil {
+		err = setupStatsTable(ctx, sh.sharedDB, "shared")
+		if err != nil {
+			return fmt.Errorf("failed to setup shared stats table: %w", err)
+		}
+		err = sh.setupLogicalReplication(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to setup logical replication: %w", err)
+		}
+	}
+	return nil
+
+}
+
+func setupStatsTable(ctx context.Context, db *sql.DB, defaultDbName string) error {
+	tx, err := db.Begin()
+	if err != nil {
+		return err
+	}
+	sqlStatement := fmt.Sprintf(`create table if not exists "rsources_stats" (
+		db_name text not null default '%s',
+		job_run_id text not null,
+		task_run_id text not null,
+		source_id text not null,
+		destination_id text not null,
+		in_count integer not null default 0,
+		out_count integer not null default 0,
+		failed_count integer not null default 0,
+		ts timestamp not null default NOW(),
+		primary key (db_name, job_run_id, task_run_id, source_id, destination_id)
+	)`, defaultDbName)
+	_, err = tx.ExecContext(ctx, sqlStatement)
+	if err != nil {
+		_ = tx.Rollback()
+		return err
+	}
+	_, err = tx.ExecContext(ctx, `create index if not exists rsources_stats_job_run_id_idx on "rsources_stats" (job_run_id)`)
+	if err != nil {
+		_ = tx.Rollback()
+		return err
+	}
+	return tx.Commit()
+}
+
+func (sh *sourcesHandler) setupLogicalReplication(ctx context.Context) error {
+
+	publicationQuery := `CREATE PUBLICATION "rsources_stats_pub" FOR TABLE rsources_stats`
+	_, err := sh.localDB.ExecContext(ctx, publicationQuery)
+	if err != nil {
+		pqError, ok := err.(*pq.Error)
+		if !ok || pqError.Code != pq.ErrorCode("42710") { // duplicate
+			return fmt.Errorf("failed to create publication on local database: %w", err)
+		}
```

Comment:
Yes, this scenario is already verified by [this test](https://github.com/rudderlabs/rudder-server/blob/70ca825e45a13d796a5ed53029b129034a38b2c1/services/rsources/handler_test.go#L407). 

More specifically, after having successfully created two local publications and 2 remote subscriptions, we create a new set of services, i.e. repeat the same steps only this time subscriptions and publications already exist, thus error code `42710` is thrown by postgres.

Reply:
Right, I saw codecov complaining about not being covered though :thinking: Why do you think that is?

Reply:
will check again, but which line isn't covered? 233 is throwing an error other than 42710

Reply:
line 245 can be covered, but line 229 is not easy to cover since a publication can be created even with an unsupported wal_level, so not easy to trigger an error there.

Reply:
Added two more tests for exceptional scenarios, both `pq.ErrorCode("42710")` are covered.

The only branches that are not covered are:
1. Failure of casting the error to a`*pq.Error`
2. Postgres error during `CREATE PUBLICATION` other than `42710`

These branches are not expected to happen when database is healthy and anyway even if they happen we will stop and return the error immediately, i.e. fail to startup the server so the pod will keep restarting, until database is healthy again.

Reply:
Thanks for checking, makes sense :+1: